File: haskell/day2/Main.hs
import Data.List (group, sort, tails)


main :: IO ()
main = do
    ids <- lines <$> readFile "input.txt"
    putStrLn ("Part one: " ++ show (partOne ids))
    putStrLn ("Part two: " ++ show (partTwo ids))


partOne :: [String] -> Int
partOne ids = length idsWithDoublettes * length idsWithTriplets
  where
    idsWithDoublettes = filter (any ((== 2) . snd)) histograms
    idsWithTriplets = filter (any ((== 3) . snd)) histograms
    histograms = map histogram ids


partTwo :: [String] -> String
partTwo ids = [a | (a,b) <- zip idA idB, a == b]
  where
    (idA,idB) = head . filter (null . drop 1 . differences) . pairs $ ids
    differences (as,bs) = filter (uncurry (/=)) (zip as bs)


-- | Counts how often each element in the input list occurs
histogram :: Ord a => [a] -> [(a, Int)]
histogram = map (\g -> (head g, length g)) . group . sort


-- | Yields all ways to pick two elements from a list (ignoring ordering)
pairs :: [a] -> [(a,a)]
pairs xs = [(a,b) | (a:as) <- tails xs, b <- as]

File: rust/day2/src/main.rs
extern crate counter;

use std::io::BufRead;
use std::io::BufReader;
use std::fs::File;
use counter::Counter;


fn part_one(ids: &Vec<String>) -> usize {
    let histograms: Vec<Counter<_>> = ids.iter().map(|id| id.chars().collect()).collect();

    let doublettes = histograms.iter().filter(
        |&h| h.values().any(|&value| value == 2)
    );

    let triplets = histograms.iter().filter(
        |&h| h.values().any(|&value| value == 3)
    );

    &doublettes.count() * &triplets.count()
}


fn pairs<T>(v: &Vec<T>) -> Vec<(&T,&T)> {
    let mut result = vec![];
    for (i, x) in v.iter().enumerate() {
        for y in v[i+1..].iter() {
            result.push((x,y));
        }
    }
    result
}


fn part_two(ids: &Vec<String>) -> String
{
    for (a,b) in pairs(ids) {
        let pairwise_chars: Vec<(char,char)> = a.chars().zip(b.chars()).collect();

        let different_chars = pairwise_chars.iter().filter(|(x,y)| x != y);
        if different_chars.count() == 1 {
            return pairwise_chars.iter().filter(|(x,y)| x == y).map(|(x,y)| x).collect();
        }
    }
    String::new()
}


fn main() {
    let input_file = File::open("input.txt").unwrap();

    let ids = BufReader::new(&input_file)
                .lines()
                .map(|line| line.unwrap())
                .collect();

    println!("Part one: {:?}", part_one(&ids));
    println!("Part two: {:?}", part_two(&ids));
}
